src/JustSaying/Fluent/PublishConfig/DynamicAddressMessagePublisher.cs

```diff
@@ -0,0 +1,112 @@
+using System.Collections.Concurrent;
+using JustSaying.Messaging;
+using JustSaying.Messaging.Interrogation;
+using JustSaying.Models;
+using Microsoft.Extensions.Logging;
+
+namespace JustSaying.Fluent;
+
+internal sealed class DynamicAddressMessagePublisher(
+    string topicArnTemplate,
+    Func<string, Message, string> topicAddressCustomizer,
+    Func<string, StaticAddressPublicationConfiguration> staticConfigBuilder,
+    ILoggerFactory loggerFactory) : IMessagePublisher, IMessageBatchPublisher
+{
+    private readonly string _topicArnTemplate = topicArnTemplate;
+    private readonly ConcurrentDictionary<string, IMessagePublisher> _publisherCache = new();
+    private readonly ConcurrentDictionary<string, IMessageBatchPublisher> _batchPublisherCache = new();
+    private readonly ConcurrentDictionary<string, SemaphoreSlim> _topicCreationLocks = new();
```

Comment:
Are we OK with these never being disposed?

Reply:
I think given Publishers tend to last the lifetime of the application, it should be okay.
To make sure unmanaged resources get cleaned up, SemaphoreSlim also (transiently) uses a SafeHandle, which will also cleanup on finalize.